sds_in_a_box/SDSCode/opensearch_utils/payload.py

```diff
@@ -0,0 +1,80 @@
+import json
+from sds_in_a_box.SDSCode.opensearch_utils.document import Document
+from opensearchpy import OpenSearch, RequestsHttpConnection
+
+
+class Payload():
+    """
+    Class to represent an OpenSearch bulk document payload.
+
+    ...
+
+
+    Attributes
+    ----------
+    payload_contents: list
+        list of json strings representing the full payload contents,
+        broken up into a list to avoid request limits when sending to OpenSearch.
+
+    Methods
+    -------
+    add_documents(documents):
+        adds document(s) to the payload to prepare for a 
+        bulk upload.
+    get_contents():
+        returns the full payload contents as a string.
+    """
+    def __init__(self):
+        self.payload_contents = []
+
+    def add_documents(self, documents):
+        """
+        Add document(s) to the payload for a bulk upload.
+        
+        Parameters
+        ----------
+        documents: Document, list of Documents
+            document(s) to be added to the payload in preparation for a bulk upload.
+        """       
+        if Document.is_document(documents):
+            self.__add_to_payload(documents)
+
+        elif type(documents) is list:
+            # check that all the objects in documents are of type Document
+            if all(Document.is_document(doc) for doc in documents): 
+                for doc in documents:
+                    self.__add_to_payload(doc)
+            
+            else:
+                raise TypeError("Document list contained at least one object that was not type Document")
+
+        else:
+            raise TypeError("Input was type {} must be type Document or list of Documents.".format(type(documents)))
+
+    def get_contents(self):
+        """Returns the contents of the payload as a string."""
+        full_contents = "".join(self.payload_contents)
+        return full_contents
+
+    def __repr__(self):
+        return str(self.payload_contents)
+
+    def __add_to_payload(self, document):
+        # TODO: not sure what the actual request limit is or how it's 
```

Comment:
Can you clarify what limit is being hit, how, and what errors you're receiving?

Reply:
When I use the opensearch-py bulk() method I get a "TransportError(429, 'Too Many Requests') error when the bulk upload is too large. The purpose of the chunking is to avoid that limit, but I've been having trouble finding official documentation on what exactly that limit is. Right now I'm using a rough string size limit that I found from iteratively increasing the chunk size until I got the error. It seems to allow me to consistently avoid getting that error, but I am still trying to figure out what that limit officially is. I originally thought it was a size issue, but after looking into it more, it seems like it's more of a rate issue, so I'm not sure why lowering the chunk size helps. Let me know if you know of anything on Libera that was used to avoid this issue.

Reply:
From Gavin based on our testing with OpenSearch:

It sounds very similar to what we did for Libera ITDC. My understanding is that 429 means there are too many simultaneous bulk upload requests for the server to handle – i.e. the ingest queue is FULL. Basically the upshot is that data is being sent faster than it can be indexed. I solved this by adding a random 1-3s delay between bulk request retries to give opensearch an opportunity to clear the queue without constantly pelting it with HTTP requests. The other error that occurred a lot during my development is a timeout error, which just means the server doesn’t have the resources to send back a response to your HTTP request at all HOWEVER, this doesn’t necessarily mean that the data didn’t get ingested, so if you retry after that, expect to either create duplicate records or get a bunch of errors that the records all exist.
 
*As we discovered in our data lab, the hands down most reliable way to fix all of this is to just throw some more ingest nodes into the cluster.* We are using m6g.large.search

Reply:
Did you upgrade to m6g.large.search to avoid needing the delay or is that delay still needed?